### ISSUE TITLE
runner with multiple generators: make it more CSP-like

### ISSUE THREAD

**Issue by getify** [getify/asynquence#34]
Right now, the `runner(..)` plugin uses a dummy-simple round-robbin scheduling to "cooperate" between multiple generators. It also uses a dummy-simple messaging mechanism to pass along one yielded value to the next generator and back, etc.

A more sophisticated mechanism is called for. One more like (though not necessarily identical to) the one detailed here: http://swannodette.github.io/2013/08/24/es6-generators-and-csp/

I'm envisioning something like:

``` js
// just a silly number generator that's promise-async
function getVal() {
   return new Promise( function(resolve,reject){
      setTimeout( function(){
         if (!getVal.__number) getVal.__number = 0;
         resolve( getVal.__number += 10 );
      }, 100 );
   } );
}

ASQ( 2, 3 )
.runner(
   // co-rountine 1
   function*(token){
      token.messages; // [2, 3]

      token.messages[0] += yield getVal(); // pauses while promise resolves
      token.messages[1] += yield getVal(); // pauses while promise resolves
      token.messages.push( token.messages[0] + token.messages[1] );

      token.messages; // [12, 23, 35]

      yield token; // hand control over to next co-rountine

      token.messages; // [ 100 ]

      token.messages[0] += yield getVal(); // pauses while promise resolves

      // implicit finish means transfer control to next co-rountine, if any
   },
   // co-rountine 2
   function*(token){
      token.messages; // [12, 23, 35]

      token.messages = [
         token.messages[0] + token.messages[1] + token.messages[2]
      ];
      token.messages; // [ 70 ]

      token.messages[0] += yield getVal(); // pauses while promise resolves
      token.messages; // [ 100 ]

      yield token; // hand control over to next co-rountine

      token.messages; // [ 140 ]

      token.messages.push( "Hello World" );

      // no magic: explicitly yield `token.messages`, or any other direct value
      // if desired. if you don't explicitly yield something non-undefined at the
      // end, the last non-`undefined`, non-`token` value yielded will be the
      // outbound message(s) from this run
      yield token.messages;
   }
)
.val( function(msgs){
   console.log( msg[0], msgs[1] ); // 140 "Hello World"
} );
```

OK, so as I wrote that out, I kinda like it even more.

Basically each co-rountine (aka generator) is given a `token` with a `messages` "channel" (array). You can add/remove from the channel as you see fit. You can even yield your current generator with a promise and have _it_ restarted as many times as you see fit. If however you actually `yield token` (or something that results in it), then you are explicitly transferring control to the next co-rountine. Ordering of control is still simple round-robbin.

Thoughts?

Would be especially interested to hear any reactions/criticisms from @swannodette. :)
